### PR TITLE
Improve code quality subscriber new action

### PR DESCRIPTION
**Pull request by arnoudhgz**
+ Refactor `else` block for returning success message to private function
+ Remove not thrown `@thrown` tag and add missing `@thrown` tag
+ Use import statements for classes
+ Because imports are being used an `if` statement can be written on one line
+ Use `addExceptionMessage` and `addSuccessMessage` instead of their deprecated counterparts

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @arnoudhgz. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 1]
```diff
@@ -125,27 +131,26 @@ public function execute()
...
                     );
                 }
 
                 $status = $this->_subscriberFactory->create()->subscribe($email);
-                if ($status == \Magento\Newsletter\Model\Subscriber::STATUS_NOT_ACTIVE) {
-                    $this->messageManager->addSuccess(__('The confirmation request has been sent.'));
-                } else {
-                    $this->messageManager->addSuccess(__('Thank you for your subscription.'));
+                if ($status == Subscriber::STATUS_NOT_ACTIVE) {
+                    $this->messageManager->addSuccessMessage(__('The confirmation request has been sent.'));
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: In this case redirect will not be executed. Currently this change adds regression. It's better to revert it or suggest another solution

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 2]
```diff
@@ -98,20 +104,20 @@ protected function validateGuestSubscription()
...
-     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws LocalizedException
+     * @throws \Zend_Validate_Exception
      * @return void
      */
     protected function validateEmailFormat($email)
     {
-        if (!\Zend_Validate::is($email, \Magento\Framework\Validator\EmailAddress::class)) {
-            throw new \Magento\Framework\Exception\LocalizedException(__('Please enter a valid email address.'));
+        if (!(new \Zend_Validate)->is($email, EmailAddress::class)) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: In all the places Magento uses Zend_Validate::is(), but in this case you suggest to replace it with object creation.
https://github.com/magento/magento2/search?p=3&q=Zend_Validate&unscoped_q=Zend_Validate
In general I agree that using static methods not the best practice, but it's better to change it in all places, also not sure if this variant is really optimal.
If you would like to propose replacement to object creation - feel free to create separate PR, but in scope of this PR - please revert this change

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php (hunk 2 above)
Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Thank you for the feedback, I only changed it because Codacy mentioned it as a new issue from my commit (because of the import statements).

Comment (ihor-sviziev) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: We're using codaxy just as additional information, not all issues we want to be fixed. As example this one :)

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 3]
```diff
@@ -125,29 +132,38 @@ public function execute()
...
 
-                $status = $this->_subscriberFactory->create()->subscribe($email);
-                if ($status == \Magento\Newsletter\Model\Subscriber::STATUS_NOT_ACTIVE) {
-                    $this->messageManager->addSuccess(__('The confirmation request has been sent.'));
-                } else {
-                    $this->messageManager->addSuccess(__('Thank you for your subscription.'));
-                }
-            } catch (\Magento\Framework\Exception\LocalizedException $e) {
-                $this->messageManager->addException(
+                $this->messageManager->addSuccessMessage($this->getSuccessMessage(
```
Comment (arnoudhgz) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: @ihor-sviziev I refactored it slightly different, using a private function instead of the if/else block

Comment (ihor-sviziev) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: You're doing few thing at once, it's hard to understand what's going on there. Previously is was more simple to understand.

Please add following code as separate line (as it was before)
```php
$status = $this->_subscriberFactory->create()->subscribe($email);
```

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 4]
```diff
@@ -125,29 +132,37 @@ public function execute()
...
             }
         }
         $this->getResponse()->setRedirect($this->_redirect->getRedirectUrl());
     }
+
+    /**
+     * @param int $status
+     * @return Phrase
+     */
+    private function getSuccessMessage($status)
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Can we define parameter and return types there?

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 5]
```diff
@@ -125,29 +132,37 @@ public function execute()
                 $this->validateEmailAvailable($email);
 
                 $subscriber = $this->_subscriberFactory->create()->loadByEmail($email);
-                if ($subscriber->getId()
-                    && $subscriber->getSubscriberStatus() == \Magento\Newsletter\Model\Subscriber::STATUS_SUBSCRIBED
-                ) {
-                    throw new \Magento\Framework\Exception\LocalizedException(
+                if ($subscriber->getId() && $subscriber->getSubscriberStatus() == Subscriber::STATUS_SUBSCRIBED) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: can we use strict comparison there? maybe just convert to expected type before?

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 6]
```diff
@@ -125,29 +132,37 @@ public function execute()
...
         $this->getResponse()->setRedirect($this->_redirect->getRedirectUrl());
     }
+
+    /**
+     * @param int $status
+     * @return Phrase
+     */
+    private function getSuccessMessage($status)
+    {
+        if ($status == Subscriber::STATUS_NOT_ACTIVE) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Can we use strict comparison there?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php (hunk 3 above)
Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: @ihor-sviziev  I was indeed doubting about this one, the old way is indeed more understandable. :+1:

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 7]
```diff
@@ -98,23 +105,23 @@ protected function validateGuestSubscription()
      * Validates the format of the email address
      *
      * @param string $email
-     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws LocalizedException
+     * @throws \Zend_Validate_Exception
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Looks like this type of exception should not be thrown. Am I wrong?

**Comment by ihor-sviziev** [comment]
Let's wait for Travis build results

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php (hunk 7 above)
Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: I got the message from PHPStorm that not all the necessary `@throws` tags were set, so I added this with the autocomplete in PHPStorm.

Comment (ihor-sviziev) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: AFAIK `\Zend_Validate_Exception` could be thrown in case if validator wasn't found. So this method actually can't throw such exceptions.
Please remove it.
Everything else looks good.

Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: In **vendor/magento/zendframework1/library/Zend/Validate.php** in the static `is` method, not all exceptions are being caught. So I should leave the `@throws` tag or write a `try/catch`.

Please see: https://github.com/magento/zf1/blob/48ac5e38c591afefb95e0d7127d00fe52d0fbb0f/library/Zend/Validate.php#L195

Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: @ihor-sviziev I found out that Magento actually had an email validator class, I required this one through dependency injection. This saves some reflection :)

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 8]
```diff
@@ -10,19 +10,31 @@
 use Magento\Customer\Model\Session;
 use Magento\Customer\Model\Url as CustomerUrl;
 use Magento\Framework\App\Action\Context;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Phrase;
+use Magento\Framework\Validator\EmailAddress;
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: I think it's better to use alias `EmailAddressValidator` or `EmailValidator` there to be clear

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 9]
```diff
@@ -32,16 +44,19 @@ class NewAction extends \Magento\Newsletter\Controller\Subscriber
...
      */
     public function __construct(
         Context $context,
         SubscriberFactory $subscriberFactory,
         Session $customerSession,
         StoreManagerInterface $storeManager,
         CustomerUrl $customerUrl,
-        CustomerAccountManagement $customerAccountManagement
+        CustomerAccountManagement $customerAccountManagement,
+        EmailAddress $emailValidator
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: According to https://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/ 
 Please make this parameter optional and get object from ObjectManager as fallback.

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<?php
+<?php declare(strict_types=1);
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Please revert this change. This changes changes behavior. For now we can declare strict types =1 only for new files

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php [hunk 11]
```diff
@@ -55,16 +70,16 @@ public function __construct(
      * Validates that the email address isn't being used by a different account.
      *
      * @param string $email
-     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws LocalizedException
      * @return void
      */
-    protected function validateEmailAvailable($email)
+    protected function validateEmailAvailable(string $email): void
```
Comment (ihor-sviziev) [inline app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Please revert adding types of arguments and return types for **all methods that were existing there**. It's signature change, that is not backward compatible

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php (hunk 10 above)
Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: @ihor-sviziev in another PR of mine, the reviewer requested to enable strict types on an existing class: https://github.com/magento/magento2/pull/16571#issuecomment-403402670

Now I am confused :(

Comment (ihor-sviziev) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: Ok. Sorry for confusing you. I will discuss it with Maintainers and we'll decide what to do in such cases. For now no need to change strict types declaration and arguments for methods, but need to fix optional parameter in constructor

Comment (ihor-sviziev) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: After discussion - we decided that we can't use strict types for existing classes as it's not backward compatible change. The same for method arguments and return types in methods that already exists.
https://github.com/magento/magento2/pull/16571#issuecomment-412021296

Please adjust your code. 
Thx

Comment (arnoudhgz) [reply app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php]: @ihor-sviziev  thank you for discussing it, at least it is now clear for me :)

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-2788](https://jira.corp.magento.com/browse/ENGCOM-2788) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
Hi @arnoudhgz. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.0. 
Please check the release notes for final confirmation.
